Filled `Builder` trait with `new_...` methods for all types

desaikd opened the pull request:
*Issue #204*

*Description of changes:*
This PR works on filling in all `new_...` methods for `Builder` trait created by #199.

*Changes:*
- Added `new_...` method for all types in Builder trait
- Modified tests to use `new_...` method construct element.

*TODO:*
- Add `From<X>` for all types in `Element` trait as defined below:
```rust
trait Element 
where
     Self:From<i64> + From<f64> + … {
```
- Add generic test to verify above mentioned element constructions `e1` and `e2`.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

almann commented on src/value/borrowed.rs:
```diff
@@ -141,12 +141,46 @@ impl<'val> SymbolToken for BorrowedSymbolToken<'val> {
+        BorrowedValue::Boolean(bool).into()
+    }
+
+    fn new_string(str: &'static str) -> Self::Element {
+        BorrowedValue::String(str).into()
+    }
```
I am not sure about this--are you able to tie the life to `'val`? If not (e.g. this requires GAT), then I would eliminate it and only generically provide the `String` based constructor (similarly for `clob` and `blob`).

almann commented on src/value/mod.rs:
```diff
@@ -487,11 +487,37 @@ pub trait Struct {
+    /// Build a `f64` using Builder
+    fn new_f64(float: f64) -> Self::Element;
+
     /// Build a `clob` using Builder
     fn new_clob(bytes: &'static [u8]) -> Self::Element;
 
```
I somehow missed this in an earlier commit--this should not be `'static` for sure, if we cannot have an appropriate lifetime due to GAT, then let's just use the owned type (e.g. `Vec<u8>`)

almann commented on src/value/mod.rs:
```diff
@@ -503,4 +529,7 @@ pub trait Builder {
     fn new_sexp(seq: Self::Sequence) -> Self::Element;
+
+    /// Build a `struct` from Struct using Builder
+    fn new_struct(struct_: Self::Struct) -> Self::Element;
```
I would change these to just spell it out (see also the implementations).

```suggestion
    fn new_struct(structure: Self::Struct) -> Self::Element;
```

almann reviewed (changes requested):
I think the builder APIs that take `&'static [u8]` and `&'static str` should be changed to their owned counterparts if you cannot tag the lifetimes correctly (ugh, GAT).

I do think this change implies changing the`Borrowed...` types to use [`Cow<'val, X>`][1].  #135 alludes to this, but it is fine if we just do it now.

[1]: https://doc.rust-lang.org/std/borrow/enum.Cow.html

desaikd commented:
Changes with https://github.com/amzn/ion-rust/pull/206/commits/83ac1b2f6b0c9a1ade51ec81392a92a26290c437:

- Added `From<X>` for scalar types in `Element`.
- Added generic scalar tests for `Element`.

TODO:

- Add generic tests for container types and `SymbolToken`.
- Add `From<X>` for container types and `SymbolToken`

almann commented on src/value/mod.rs:
```diff
@@ -258,10 +259,13 @@ impl Eq for AnyInt {}
-    type SymbolToken: SymbolToken + ?Sized;
-    type Sequence: Sequence<Element = Self> + ?Sized;
-    type Struct: Struct<FieldName = Self::SymbolToken, Element = Self> + ?Sized;
+pub trait Element
+where
+    Self: From<i64> + From<bool> + From<Decimal> + From<Timestamp> + From<f64> + From<BigInt>,
+{
+    type SymbolToken: SymbolToken + ?Sized + Debug + PartialEq;
+    type Sequence: Sequence<Element = Self> + ?Sized + Debug + PartialEq;
+    type Struct: Struct<FieldName = Self::SymbolToken, Element = Self> + ?Sized + Debug + PartialEq;
```
Since you added `Debug` to all of the downstream traits, should `Element` have `Self: Debug`?

almann reviewed (commented):
Looks good overall--I like how the generic tests worked out, should we eliminate the non-generic tests now (I would run coverage before and after removal to be sure that we still are getting everything).

almann commented on src/value/owned.rs:
```diff
@@ -666,20 +719,20 @@ mod value_tests {
     #[rstest(
         elem, ion_type, valid_ops_iter, op_assert,
         case::null(
-            IonType::Null.into(),
+            OwnedElement::new_null(IonType::Null),
             IonType::Null,
             IsNull,
             &|e: &OwnedElement| assert_eq!(true, e.is_null())
         ),
         // TODO more null testing (probably its own fixture)
```
Do these tests matter now that you have generic ones?

desaikd replied:
I meant to do this in separate PR as it would require to add more tests in generic test suite.

almann commented on src/value/borrowed.rs:
```diff
@@ -664,20 +718,20 @@ mod borrowed_value_tests {
     #[rstest(
         elem, ion_type, valid_ops_iter, op_assert,
         case::null(
-            IonType::Null.into(),
+            BorrowedElement::new_null(IonType::Null),
             IonType::Null,
             IsNull,
             &|e: &BorrowedElement| assert_eq!(true, e.is_null())
         ),
         // TODO more null testing (probably its own fixture)
         case::bool(
-            true.into(),
+            BorrowedElement::new_bool(true),
```
Do these tests matter now that you have generic ones?

almann reviewed (commented):
I think this is pretty close.  Do we really need the owned/borrowed tests now that you've made everything generic?

desaikd commented on src/value/mod.rs:
```diff
@@ -258,10 +259,13 @@ impl Eq for AnyInt {}
-    type Struct: Struct<FieldName = Self::SymbolToken, Element = Self> + ?Sized;
+pub trait Element
+where
+    Self: From<i64> + From<bool> + From<Decimal> + From<Timestamp> + From<f64> + From<BigInt>,
```
It does seem like a reasonable bound to me as we are not taking `From<AnyInt>` as the bound here.  Did you mean to have something else in mind for `BigInt`?

almann commented on src/value/borrowed.rs:
```diff
@@ -70,20 +69,17 @@ impl<'val> BorrowedSymbolToken<'val> {
         self.text = Some(text);
         self
     }
 
-    /// Decorates the [`BorrowedSymbolToken`] with a local ID.
-    pub fn with_local_sid(mut self, local_sid: SymbolId) -> Self {
+    fn with_local_sid_borrrowed(mut self, local_sid: SymbolId) -> Self {
         self.local_sid = Some(local_sid);
         self
     }
 
-    /// Decorates the [`BorrowedSymbolToken`] with an [`BorrowedImportSource`].
-    pub fn with_source(mut self, table: &'val str, sid: SymbolId) -> Self {
+    fn with_source_borrowed(mut self, table: &'static str, sid: SymbolId) -> Self {
```
Why do we have `..._borrowed` and the forms below that don't have the suffix?

desaikd replied:
changed this to construct a new `SymbolToken ` for all the `with _...` construction methods.

almann commented on src/value/borrowed.rs:
```diff
@@ -137,6 +133,18 @@ impl<'val> SymbolToken for BorrowedSymbolToken<'val> {
         self.source.as_ref()
     }
+
+    fn with_text(self, text: &'static str) -> Self {
+        self.with_text_borrowed(text)
+    }
+
+    fn with_local_sid(self, local_sid: SymbolId) -> Self {
+        self.with_local_sid_borrrowed(local_sid)
+    }
+
+    fn with_source(self, table: &'static str, sid: SymbolId) -> Self {
+        self.with_source_borrowed(table, sid)
+    }
```
Let's make sure we have an issue to deal with this `'static` lifetime problem the APIs will be very marginal with this...

almann commented on src/value/owned.rs:
```diff
@@ -71,20 +71,17 @@ impl OwnedSymbolToken {
+        self.text = Some(Rc::from(text));
         self
     }
 
-    /// Decorates the [`OwnedSymbolToken`] with a local ID.
-    pub fn with_local_sid(mut self, local_sid: SymbolId) -> Self {
+    fn with_local_sid_owned(mut self, local_sid: SymbolId) -> Self {
         self.local_sid = Some(local_sid);
         self
     }
 
-    /// Decorates the [`OwnedSymbolToken`] with an [`OwnedImportSource`].
-    pub fn with_source<T: Into<Rc<str>>>(mut self, table: T, sid: SymbolId) -> Self {
+    fn with_source_owned(mut self, table: &'static str, sid: SymbolId) -> Self {
         self.source = Some(OwnedImportSource::new(table, sid));
```
Same question on this API, why do we need `..._owned` variants when they look the same as the non-suffix ones below?

almann reviewed (commented):
I have some questions around the API factoring below.

almann reviewed (approved):
As long as we have an issue around our `'static` lifetimes and an issue (or its been removed) for the redundant tests, :rocket:.
